### PR TITLE
[RO-53] Dedicated LTI instance, line item and assignment ingestion commands

### PR DESCRIPTION
**Pull request by Skyld**
Ticket: https://oat-sa.atlassian.net/browse/RO-53

### Added
- Added [roster:ingest:lti-instance](docs/cli/lti-instance-ingester-command.md) command.
- Added [roster:ingest:line-item](docs/cli/line-item-ingester-command.md) command.
- Added [roster:ingest:user](docs/cli/user-ingester-command.md) command.
- Added [roster:ingest:assignment](docs/cli/assignment-ingester-command.md) command.
- Added possibility to use multiple filesystem instances with the help of [Storage registry](docs/storage-registry.md).

### Removed
- Removed `roster:ingest` command.
- Removed `roster:assignments:bulk-create` command.
- Removed `roster:assignments:bulk-cancel` command.

### CODE REVIEW

File: src/DataTransferObject/AssignmentDtoCollection.php [hunk 1]
```diff
@@ -29,39 +29,57 @@
...
+                    return $assignmentDto->getUsername();
+                },
+                $this->assignments
+            )
+        );
+    }
+
+    public function isEmpty(): bool
+    {
+        return count($this) === 0;
```
Comment (luisfn) [inline src/DataTransferObject/AssignmentDtoCollection.php:70]: Should not count $this->assignments?

**Review by luisfn — CHANGES_REQUESTED** [review]
Thanks for the great work on this PR. Ingestion process seems much better and concise now.
Didn't found any big issue, please just check the comments.

File: src/Command/Ingester/LtiInstanceIngesterCommand.php [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                    0,
+                    $rawLtiInstance['label'],
+                    $rawLtiInstance['ltiLink'],
+                    $rawLtiInstance['ltiKey'],
+                    $rawLtiInstance['ltiSecret']
+                );
+
+                $this->ltiInstanceRepository->persist($ltiInstance);
+
+                if ($numberOfProcessedRows % $this->batchSize === 0) {
```
Comment (andreluizmachado) [inline src/Command/Ingester/LtiInstanceIngesterCommand.php:98]: Can we isolate this conditional? The reason is that it is showing up several times and to better explain what it means.

File: src/Command/Ingester/LtiInstanceIngesterCommand.php [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+            $numberOfProcessedRows = 0;
+            $persisted = false;
+            foreach ($this->csvReader->getRecords() as $rawLtiInstance) {
+                $this->validateRow($rawLtiInstance, 'label', 'ltiLink', 'ltiKey', 'ltiSecret');
+
+                $numberOfProcessedRows++;
+                $persisted = false;
+
+                $ltiInstance = new LtiInstance(
```
Comment (andreluizmachado) [inline src/Command/Ingester/LtiInstanceIngesterCommand.php:88]: As you did in the other commands, you could create a factory method :)

File: src/Command/Ingester/LtiInstanceIngesterCommand.php [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    {
+        parent::initialize($input, $output);
+
+        $this->symfonyStyle->title('Simple Roster - LTI Instance Ingester');
+    }
+
+    /**
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    protected function execute(InputInterface $input, OutputInterface $output): int
```
Comment (andreluizmachado) [inline src/Command/Ingester/LtiInstanceIngesterCommand.php:72]: I might be wrong, but we could put this method also in the Abstract class. EX:
```
protected function execute(InputInterface $input, OutputInterface $output): int
     {
         $this->symfonyStyle->text($this->getStartComment());
         $this->symfonyStyle->newLine();

          try {
             $this->progressBar->start();

              $numberOfProcessedRows = 0;
             $persisted = false;
             foreach ($this->csvReader->getRecords() as $rawLineItem) {
                 $this->validateRow(
                     $rawLineItem,
                     ...$this->getCsvColumns()
                 );

                  $numberOfProcessedRows++;
                 $persisted = false;

                  $this->processRow($rawLineItem);

                  if ($numberOfProcessedRows % $this->batchSize === 0) {
                     if (!$this->isDryRun) {
                         $this->processBatch();
                     }

                      $persisted = true;

                      $this->progressBar->advance($this->batchSize);
                 }
             }

              if (!$this->isDryRun && !$persisted) {
                 $this-> processBatch();
             }

              $this->progressBar->finish();

              $this->symfonyStyle->newLine(2);

              $verificationCommentMessage = $this->getVerificationComment();

              if ($this->isDryRun) {
                 $this->symfonyStyle->warning(
                     $this->getSucessfulComment($numberOfProcessedRows)
                 );

                  $this->symfonyStyle->comment($verificationCommentMessage);

                  return 0;
             }

              $this->symfonyStyle->success(
                 $this->getSucessfulComment($numberOfProcessedRows)
             );

              $this->symfonyStyle->comment($verificationCommentMessage);
         } catch (Throwable $exception) {
             $this->symfonyStyle->error($exception->getMessage());

              return 1;
         }

          return 0;
     }
```

**Review by andreluizmachado — CHANGES_REQUESTED** [review]
Nice Job 👍 @Skyld. I just have a few comments and one refactory suggestion

File: src/Command/Ingester/AssignmentIngesterCommand.php [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    /** @var LineItemRepository */
+    private $lineItemRepository;
+
+    public function __construct(
+        CsvReaderBuilder $csvReaderBuilder,
+        StorageRegistry $storageRegistry,
+        AssignmentIngester $assignmentIngester,
+        LineItemRepository $lineItemRepository
+    ) {
+        $this->csvReaderBuilder = $csvReaderBuilder;
```
[the remark this answers is not in the document]
Comment (Skyld) [reply src/Command/Ingester/AssignmentIngesterCommand.php:54]: Not sure what do you mean :) It is passed to the parent class.

File: src/Command/Ingester/AssignmentIngesterCommand.php [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    protected function getIngesterCommandName(): string
+    {
+        return self::NAME;
+    }
+
+    protected function configure(): void
+    {
+        parent::configure();
+
+        $this->setDescription('Assignment ingestion from various sources');
```
[the remark this answers is not in the document]
Comment (Skyld) [reply src/Command/Ingester/AssignmentIngesterCommand.php:70]: I rephrased the descriptions. Available storages are listed in `storage` option description (see parent class)

File: src/DataTransferObject/AssignmentDtoCollection.php (hunk 1 above)
Comment (Skyld) [reply src/DataTransferObject/AssignmentDtoCollection.php:70]: It doesn't matter, but invoking the method is maybe a little bit nicer indeed. Changed it.

File: docs/storage-registry.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    storages:
+        local.storage:
+            adapter: 'local'
+            options:
+                directory: '%kernel.project_dir%'
+```
+
+It is pre-configured to use the project root as base directory.
+
+## How to configure an AWS S3 bucket storage
```
[the remark this answers is not in the document]
Comment (Skyld) [reply docs/storage-registry.md:26]: Mmmm... did you read the document? What you've described is what is actually described in it :) maybe it's not clear enough?

File: src/Command/Ingester/AbstractCsvIngesterCommand.php [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+            ->setDelimiter((string)$input->getOption('delimiter'))
+            ->build(
+                (string)$input->getArgument('path'),
+                (string)$input->getOption('storage')
+            );
+
+        $process = new Process(['wc', '-l', $this->csvReader->getPathname()]);
+        $process->run();
+
+        // In tests with in memory adapter file content cannot be counted with process.
```
[the remark this answers is not in the document]
Comment (Skyld) [reply src/Command/Ingester/AbstractCsvIngesterCommand.php:132]: Yes, there is no actual filesystem involved, it's in memory :)

File: src/Command/Ingester/LtiInstanceIngesterCommand.php (hunk 2 above)
Comment (Skyld) [reply src/Command/Ingester/LtiInstanceIngesterCommand.php:98]: It is not showing up several times, only once per command. I see your point about explaining it, but I don't know how to make it more clear. Any suggestions?

File: src/Command/Ingester/LtiInstanceIngesterCommand.php (hunk 3 above)
Comment (Skyld) [reply src/Command/Ingester/LtiInstanceIngesterCommand.php:88]: Good catch ,added

File: config/packages/flysystem.yaml [hunk 9]
```diff
@@ -0,0 +1,7 @@
+# Read the documentation at https://github.com/thephpleague/flysystem-bundle/blob/master/docs/1-getting-started.md
+flysystem:
+    storages:
+        local.storage:
+            adapter: 'local'
+            options:
+                directory: '%kernel.project_dir%'
```
[the remark this answers is not in the document]
Comment (Skyld) [reply config/packages/flysystem.yaml:7]: Yes, the scope of the PR was widened indeed, sorry for it 👍

File: src/Command/Ingester/LtiInstanceIngesterCommand.php (hunk 4 above)
Comment (Skyld) [reply src/Command/Ingester/LtiInstanceIngesterCommand.php:72]: The idea is good, unfortunately there are slight differences between the 4 commands if you check them, which would make the abstraction too complicated. Just one example: How do you define what to process in `processbatch()`?

File: src/Command/Ingester/AssignmentIngesterCommand.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (Skyld) [reply src/Command/Ingester/AssignmentIngesterCommand.php:54]: Ah I am blind, thanks, fixed :D

**Review by uncleempty — APPROVED** [review]
:+1:

File: src/Command/Ingester/LtiInstanceIngesterCommand.php (hunk 2 above)
Comment (Skyld) [reply src/Command/Ingester/LtiInstanceIngesterCommand.php]: Fixed

**Review by andreluizmachado — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Changelog is updated according to changes (if applicable)
- [x] Documentation is updated according to changes (if applicable)

**Pull request merged**